File: rtsp/src/main/java/com/pedro/rtsp/rtcp/SenderReportTcp.java
package com.pedro.rtsp.rtcp;

import android.util.Log;
import com.pedro.rtsp.rtsp.RtpFrame;
import com.pedro.rtsp.utils.RtpConstants;
import java.io.IOException;
import java.io.OutputStream;

/**
 * Created by pedro on 8/11/18.
 */

public class SenderReportTcp extends BaseSenderReport {

  private OutputStream outputStream;
  private byte tcpHeader[];

  public SenderReportTcp() {
    super();
    tcpHeader = new byte[] { '$', 0, 0, PACKET_LENGTH };
  }

  @Override
  public void setDataStream(OutputStream outputStream, String host) {
    this.outputStream = outputStream;
  }

  @Override
  public void sendReport(byte[] buffer, RtpFrame rtpFrame, String type, int packetCount,
      int octetCount, boolean isEnableLogs) throws IOException {
    sendReportTCP(buffer, rtpFrame.getChannelIdentifier(), type, packetCount, octetCount,
        isEnableLogs);
  }

  @Override
  public void close() {

  }

  private void sendReportTCP(byte[] buffer, byte channelIdentifier, String type, int packet,
      int octet, boolean isEnableLogs) throws IOException {
    synchronized (RtpConstants.lock) {
      tcpHeader[1] = (byte) (channelIdentifier + 1);
      outputStream.write(tcpHeader);
      outputStream.write(buffer, 0, PACKET_LENGTH);
      outputStream.flush();
      if (isEnableLogs) {
        Log.i(TAG, "wrote report: " + type + ", packets: " + packet + ", octet: " + octet);
      }
    }
  }
}

File: rtsp/src/main/java/com/pedro/rtsp/rtcp/SenderReportTcp.kt
/*
 * Copyright (C) 2024 pedroSG94.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pedro.rtsp.rtcp

import com.pedro.common.socket.base.TcpStreamSocket
import com.pedro.rtsp.rtsp.RtpFrame
import com.pedro.rtsp.utils.RtpConstants
import java.io.IOException

/**
 * Created by pedro on 8/11/18.
 */
class SenderReportTcp : BaseSenderReport() {

  private var socket: TcpStreamSocket? = null
  private val tcpHeader: ByteArray = byteArrayOf('$'.code.toByte(), 0, 0, RtpConstants.REPORT_PACKET_LENGTH.toByte())

  @Throws(IOException::class)
  override suspend fun setSocket(socket: TcpStreamSocket) {
    this.socket = socket
  }

  @Throws(IOException::class)
  override suspend fun sendReport(buffer: ByteArray, rtpFrame: RtpFrame) {
    sendReportTCP(buffer, rtpFrame.channelIdentifier)
  }

  override suspend fun close() {}

  @Throws(IOException::class)
  private suspend fun sendReportTCP(buffer: ByteArray, channelIdentifier: Int) {
    tcpHeader[1] = (2 * channelIdentifier + 1).toByte()
    socket?.write(tcpHeader)
    socket?.write(buffer, 0, RtpConstants.REPORT_PACKET_LENGTH)
    socket?.flush()
  }
}